### ISSUE TITLE
Error in the adversarial least-squares loss for the discriminator

### ISSUE THREAD

**Issue by neel-dey** [simontomaskarlsson/CycleGAN-Keras#22]
Hi Simon, thanks for the straightforward CycleGAN port.

I believe there might be an error in your loss function coming in from the last layer of the discriminator. As you're implementing the least-squares loss, the output of the discriminator has to be real-valued and unconstrained for the theory to work, whereas you're currently running the output through a sigmoid and thus constraining the range. See here for a discussion on the official CycleGAN repo as well: https://github.com/junyanz/CycleGAN/issues/70

Practically, applying the original implementation to the maps dataset leads to a near-fp32 ~~numerical precision~~ machine epsilon discriminator cost within just a few epochs. Removing the sigmoid fixes this.

**Comment by kpagels** [comment]
I was actually wondering the same thing. 

But Im just curious @neel-dey, what is a "near-fp32 numerical precision discriminator cost"? Does that mean the discriminators error is near 0?

And is the fix just to remove "x = Activation('sigmoid')(x)"? Or should something else be added?

**Comment by neel-dey** [comment]
Ah, I meant machine epsilon and not numerical precision (I blame jetlag). _Very_ loosely, since all computations are done with single precision (which has an epsilon of ~2.24e-7) and the discriminator objective quickly approaches this value, you cannot trust any resulting computed value due to rounding errors.  You may want to flip through M. Overton's "Numerical Computing with IEEE Floating Point Arithmetic" if you're interested in this. To answer your question directly, in a sense the discriminator objective is near zero.

Yes, you can just remove that line and everything works as expected. Hope that helps.

**Comment by simontomaskarlsson** [comment]
Hi @neel-dey,
Thank you for commenting and for the precise description. I will update the model right away.

Regards,
Simon

**Comment by neel-dey** [comment]
Happy to help :)
